consumer/src/main/java/com/flipkart/varadhi/consumer/UnGroupedMessageSrc.java

```diff
@@ -0,0 +1,50 @@
+package com.flipkart.varadhi.consumer;
+
+import com.flipkart.varadhi.entities.Offset;
+import com.flipkart.varadhi.spi.services.Consumer;
+import com.flipkart.varadhi.spi.services.PolledMessage;
+import com.flipkart.varadhi.spi.services.PolledMessages;
+import lombok.RequiredArgsConstructor;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.LinkedBlockingQueue;
+
+@RequiredArgsConstructor
+public class UnGroupedMessageSrc<O extends Offset> implements MessageSrc {
+
+    final Consumer<O> consumer;
+
+    private final BlockingQueue<MessageTracker> buffer = new LinkedBlockingQueue<>();
+
+    @Override
+    public CompletableFuture<Integer> nextMessages(MessageTracker[] messages) {
+        int offset = fetchFromBuffer(messages);
+        if (offset < messages.length) {
+            return consumer.receiveAsync()
+                    .thenApply(polledMessages -> processPolledMessages(offset, polledMessages, messages));
+        }
+        return CompletableFuture.completedFuture(offset);
```

Comment:
this impl assumes that we cannot do 

```
msgSrc.nextMessages(arr_1);
msgSrc.nextMessages(arr_2); // immediately 
```

which is fine, as that is how we are going to call it. but this needs to go into the documentation.

Reply:
I think the above scenario will cause two concurrent threads to read from the buffer (which will be thread safe) followed by batchReceiveAsync on the consumer.

From the pulsar docs, they recommend that this should be called subsequently only when prev completable future completes otherwise it can create a backlog of receive requests. 

So this assumption is mostly due to the concerns from the consumer side.